## Import standard libraries

In [1]:
import torch
import torch.nn as nn  # we'll use this a lot going forward!
import torch.nn.functional as F

import numpy as np
import warnings

# Import matplotlib library and setup environment for plots
%matplotlib inline
%config InlineBackend.figure_format='retina'
from matplotlib import pyplot as plt, rc

# Import pandas and set pandas DataFrame visualization parameters
from IPython.display import display
import pandas as pd
pd.options.display.max_columns = None
pd.options.display.max_rows = None

# Set rendering parameters to use TeX font if not working on Juno app.
from pathlib import Path
import os
if not '/private/var/' in os.getcwd():
    rc('font', **{'family': 'serif', 'serif': ['Computer Modern'], 'size': 11})
    rc('text', usetex=True)
    
# Get current working directory path for the tool parent folder and print it.
parent_folder = 'Tool'
cwd = str(Path(os.getcwd()[:os.getcwd().index(parent_folder)+len(parent_folder)]))
print('Parent working directory: %s' % cwd)


Parent working directory: /Users/jjrr/Documents/SCA-Project/Tool


## Import user defined libraries

In [2]:
# Import custom libraries from local folder.
import sys
sys.path.append("..")

from library.irplib import utils, eda, sdg

## Data preparation

### Import training dataset

In [3]:
# Import transformed training dataset
df = eda.import_cdm_data(os.path.join(cwd,'data','esa-challenge','train_data_transformed.csv'))

# Count number of CDMs available per event
nb_cdms = df.groupby(['event_id']).count()['time_to_tca'].to_numpy(dtype=np.int32)

# Get information summary on TS suitable for training
min_cdms = 5
print(f'Events suitable for training: {np.sum(nb_cdms>min_cdms)}'
      f' ({np.sum(nb_cdms>min_cdms)/len(nb_cdms)*100:.2f}%)')
print(f'TSs with event_id integrity for training: {np.sum(nb_cdms[nb_cdms>min_cdms]-min_cdms)}')

# Show first data points to explore data types
display(df.head(10))

Events suitable for training: 9400 (71.46%)
TSs with event_id integrity for training: 104099


event_id  time_to_tca mission_id       risk  max_risk_estimate  \
0        0     1.566798          5 -10.204955          -7.834756   
1        0     1.207494          5 -10.355758          -7.848937   
2        0     0.952193          5 -10.345631          -7.847406   
3        0     0.579669          5 -10.337809          -7.845880   
4        0     0.257806          5 -10.391260          -7.852942   
5        1     6.530455          5  -7.561299          -7.254301   
6        1     5.561646          5  -9.315693          -7.468904   
7        1     5.226504          5  -7.422508          -7.051001   
8        1     3.570013          5  -9.248105          -7.327533   
9        2     6.983474          2 -10.816161          -6.601713   

   max_risk_scaling  miss_distance  relative_speed  relative_position_r  \
0          8.602101        14923.0         13792.0                453.8   
1          8.956374        14544.0         13792.0                474.3   
2          8.932195        14475.0         13792.0                474.6   
3          8.913444        14579.0         13792.0                472.7   
4          9.036838        14510.0         13792.0                478.7   
5          2.746782         2392.0          3434.0                 74.3   
6          7.223137         3587.0          3434.0                 99.0   
7          2.956639         7882.0          3434.0                -50.0   
8          7.425994        26899.0          3434.0                -82.0   
9         13.293159        22902.0         14348.0              -1157.6   

   relative_position_t  relative_position_n  relative_velocity_r  \
0               5976.6             -13666.8                 -7.2   
1               5821.2             -13319.8                 -7.0   
2               5796.2             -13256.1                 -7.0   
3               5838.9             -13350.7                 -7.0   
4               5811.1             -13288.0                 -7.0   
5               2317.1               -589.4                 25.9   
6               3475.4               -885.1                 24.7   
7              -7638.3               1945.7                 36.8   
8             -26067.0               6638.2                 56.8   
9              -6306.2              21986.3                 15.8   

   relative_velocity_t  relative_velocity_n t_time_lastob_start  \
0             -12637.0              -5525.9                 1.0   
1             -12637.0              -5525.9                 1.0   
2             -12637.0              -5525.9                 1.0   
3             -12637.0              -5525.9                 1.0   
4             -12637.0              -5525.9                 1.0   
5               -847.8              -3328.2                 1.0   
6               -847.8              -3328.2                 1.0   
7               -847.7              -3328.2                 1.0   
8               -847.8              -3328.2                 1.0   
9             -13792.0              -3957.1                 1.0   

  t_time_lastob_end  t_recommended_od_span  t_actual_od_span  t_obs_available  \
0               0.0                   3.78              3.78              459   
1               0.0                   3.79              3.79              456   
2               0.0                   3.79              3.80              456   
3               0.0                   3.86              3.86              443   
4               0.0                   3.86              3.86              440   
5               0.0                   4.06              4.06              432   
6               0.0                   3.91              3.91              449   
7               0.0                   3.99              3.99              434   
8               0.0                   4.04              4.04              430   
9               0.0                   3.92              3.92              444   

   t_obs_used  t_residuals_accepted  t_weighted_

### Feature engineering

#### Input vs output and categorical vs continuous input features segregation

In [4]:
# Get list of event_ids
event_ids = df['event_id'].cat.categories.values

# Initialize output features
out_features = ["risk", "max_risk_estimate"]

# Get input features
in_features = [f for f in df.columns if not f in out_features and \
               not f in ['event_id', 'mission_id', 'max_risk_scaling']]

# Separate categorical features from continuous features
cat_features = [f for f in in_features if df[f].dtype=='category']
cont_features = [f for f in in_features if not f in cat_features]

#### Continuous features: variability analysis per event

In [5]:
# Compute standard deviation, and mean of every feature per event_id.
events_std    = df[['event_id'] + cont_features].groupby(['event_id']).std()
events_mean   = df[['event_id'] + cont_features].groupby(['event_id']).mean()


# Evaluate variability of feature per event_id using the order of magnitude
events_var = pd.DataFrame(index=event_ids, columns=cont_features)

for feature in cont_features: 
    events_var[feature] = (events_mean[feature]/events_std[feature]).apply(lambda x: utils.om(x))


# Initialize and print criteria used to class features as globally constant or variable
om_criteria = 2
p_criteria = 0.1

print(f'Criteria for which a feature is considered CONSTANT:\n'
      f' (1.) - Locally (for a given event): if its mean is {om_criteria} orders of magnitude bigger than its standard deviation, or its standard deviation is 0.\n'
      f' (2.) - Globally (for the entire dataset): if criteria (1.) is met in more than {p_criteria*100:2.0f}% of the total number of events.\n')

# Get array of continuous features that are constant and variables
var_features   = [f for f in cont_features if np.sum(events_var[f]<om_criteria)/len(events_var)>p_criteria]
cons_features  = [f for f in cont_features if f not in var_features]

# Print results of analysis
print(f'Constant features: {len(cons_features)}')
for cf in cons_features: print(f' - {cf:25s} ({np.sum(events_var[cf]>=om_criteria)/len(events_var)*100:5.2f}%)')

print(f'\nVariable features: {len(var_features)}')
for vf in var_features: print(f' - {vf:25s} ({np.sum(events_var[vf]<om_criteria)/len(events_var)*100:5.2f}%)')


Criteria for which a feature is considered CONSTANT:
 (1.) - Locally (for a given event): if its mean is 2 orders of magnitude bigger than its standard deviation, or its standard deviation is 0.
 (2.) - Globally (for the entire dataset): if criteria (1.) is met in more than 10% of the total number of events.

Constant features: 19
 - relative_speed            (98.23%)
 - relative_velocity_t       (97.45%)
 - relative_velocity_n       (98.27%)
 - t_rcs_estimate            (94.15%)
 - t_j2k_sma                 (98.29%)
 - t_j2k_ecc                 (96.62%)
 - t_j2k_inc                 (98.29%)
 - c_j2k_sma                 (98.29%)
 - c_j2k_ecc                 (90.09%)
 - c_j2k_inc                 (98.29%)
 - t_span                    (98.29%)
 - c_span                    (98.29%)
 - t_h_apo                   (98.29%)
 - t_h_per                   (98.29%)
 - c_h_apo                   (98.29%)
 - c_h_per                   (98.29%)
 - geocentric_latitude       (95.96%)
 - azimuth           

## Time-Series Forecasting problem

In [6]:
def event_ts_sets(full_seq:np.ndarray, window_size:int, events_to_forecast:int=1) -> list:  
    """Get all training Time-Series sets (sequence->target) 
    associated to an event.

    Args:
        full_sequence (np.ndarray): Array containing the full sequence 
        of data for a given event.
        window_size (int): Window size of events.
        events_to_forecast (int): Number of events to forecast. Defaults to 1.

    Returns:
        list: List of tuples with sequences and labels.
    """

    # Get number of TS sets to extract from the full sequence. 
    n = len(full_seq) - (window_size + events_to_forecast)+1

    # Initialize Time-Series sets list containing tuples with sequence-target
    # for a given event.
    ts_sets = []

    # Create the list of Time-Series sets using a loop.
    for i in range(n):

        # Get sequence and target value for element i
        seq_i       = full_seq[i:i+window_size]
        target_i    = full_seq[i+window_size:i+window_size+events_to_forecast]

        # Add tuple to the output list
        ts_sets.append((seq_i, target_i))

    return ts_sets

In [14]:
from tqdm import tqdm

window_size = 5
events_to_forecast = 1

# Count number of CDMs (full Time-Series sequence) per event
ts_events  = df[['event_id', 'time_to_tca']].groupby(['event_id']).count().rename(columns={'time_to_tca':'nb_events'})

# Exclude those events that do not have a minimum number of CDMs equal to the window_size + events_to_forecast
ts_events = ts_events.drop(ts_events[ts_events['nb_events']<window_size+events_to_forecast].index)
events_filter = list(ts_events.index.values)

# Get time-series sets for every continuous variable feature 
# (constant features by definition do not need to be forecasted)
features_ts = {}
for f, feature in enumerate([f for f in in_features if f in var_features]):

    features_ts[feature] = []
    for event_id in tqdm(events_filter, desc=f'Extracting time-series for {feature:^25s} (Overall progress: {(f+1)/len(in_features)*100:5.1f}%)...'):
        full_seq = df[df['event_id']==event_id][feature].to_numpy()

        features_ts[feature] = features_ts[feature] + event_ts_sets(full_seq, window_size)




Extracting time-series for        time_to_tca        (Overall progress:   1.0%)...: 100%|██████████| 9400/9400 [00:05<00:00, 1811.99it/s]
Extracting time-series for       miss_distance       (Overall progress:   2.0%)...: 100%|██████████| 9400/9400 [00:05<00:00, 1869.21it/s]
Extracting time-series for      relative_speed       (Overall progress:   3.1%)...: 100%|██████████| 9400/9400 [00:04<00:00, 1901.12it/s]
Extracting time-series for    relative_position_r    (Overall progress:   4.1%)...: 100%|██████████| 9400/9400 [00:05<00:00, 1869.75it/s]
Extracting time-series for    relative_position_t    (Overall progress:   5.1%)...: 100%|██████████| 9400/9400 [00:04<00:00, 1896.61it/s]
Extracting time-series for    relative_position_n    (Overall progress:   6.1%)...: 100%|██████████| 9400/9400 [00:04<00:00, 1908.17it/s]
Extracting time-series for    relative_velocity_r    (Overall progress:   7.1%)...: 100%|██████████| 9400/9400 [00:04<00:00, 1931.29it/s]
Extracting time-series for    rela

In [4]:
# Get all columns from the dataset
features_groups = {
    "targets": [
        "risk", "max_risk_estimate", "max_risk_scaling"
        ],
    "ids": [
        "event_id", "mission_id"
        ],
    "conjunction": [
        "time_to_tca", "c_object_type", "miss_distance", "mahalanobis_distance", 
        "geocentric_latitude", "azimuth", "elevation",
        "F10", "F3M", "SSN", "AP"
        ],
    "relative_state": [
        "relative_position_r", "relative_position_t", "relative_position_n",
        "relative_velocity_r", "relative_velocity_t", "relative_velocity_n",
        "relative_speed"
        ],
    "objects": {
        "coefficients": [
            "cd_area_over_mass", "cr_area_over_mass"
        ],
        "covariance": [
            "ct_r", "cn_r", "cn_t",
            "crdot_r", "crdot_t", "crdot_n",
            "ctdot_r", "ctdot_t", "ctdot_n",
            "cndot_r", "cndot_t", "cndot_n",
            "ctdot_rdot", "cndot_rdot", "cndot_tdot",
            "sigma_r", "sigma_t", "sigma_n",
            "sigma_rdot", "sigma_tdot", "sigma_ndot" 
        ],
        "orb_elements": [
            "h_apo", "h_per", "j2k_ecc", "j2k_inc", "j2k_sma"
        ],
        "miscellaneous": [
            "span", "actual_od_span", "recommended_od_span", "obs_available", "obs_used",
            "time_lastob_end", "time_lastob_start", "position_covariance_det", 
            "rcs_estimate", "sedr", "residuals_accepted", "weighted_rms"
        ]
    }
}

### Converting data from Pandas DataFrame to Pytorch Tensors

In [17]:
# Initialize output features
out_features = ["risk", "max_risk_estimate"]

# Get input features
in_features = [f for f in df.columns if not f in out_features and \
               not f in ['event_id', 'mission_id']]

# Separate categorical features from continuous features
cat_features = [f for f in in_features if df[f].dtype=='category']
cont_features = [f for f in in_features if not f in cat_features]


#### Embedding categorical input features

An embedding is a vector representation of a categorical variable. The representation of this vector is computed through the use of NN models/techniques that take into account potential relation between categories in order to create the vector representation for each category.

In practice, an embedding matrix is a lookup table for a vector. Each row of an embedding matrix is a vector for a unique category.

The main advantadge of using embeddings instead of One Hot/Dummy Encoding techniques (one column per unique value of categorical feature with 0s and 1s) is that it can preserve the natural order and common relationships between the categorical features. For example, we could represent the days of the week with 4 floating-point numbers each, and two consecutive days would look more similar than two weekdays that are days apart from each other.


The rule of thumb for determining the embedding size (number of elemens per array) is to divide the number of unique entries in each column by 2, but not to exceed 50.

In [16]:
# This will set embedding sizes for the categorical columns:
# an embedding size is the length of the array into which every category
# is converted
cat_szs = [len(df[f].cat.categories) for f in cat_features]
emb_szs = [(size, min(50, (size+1)//2)) for size in cat_szs]

for f, feature in enumerate(cat_features):
    print(f'Feature {feature:20s} Embedding size: {emb_szs[f]} '
          f'(Unique vectors: {emb_szs[f][0]} | Length: {emb_szs[f][1]})')

Feature t_time_lastob_start  Embedding size: (3, 2) (Unique vectors: 3 | Length: 2)
Feature t_time_lastob_end    Embedding size: (3, 2) (Unique vectors: 3 | Length: 2)
Feature c_object_type        Embedding size: (5, 3) (Unique vectors: 5 | Length: 3)
Feature c_time_lastob_start  Embedding size: (3, 2) (Unique vectors: 3 | Length: 2)
Feature c_time_lastob_end    Embedding size: (3, 2) (Unique vectors: 3 | Length: 2)


In [ ]:
from torch.utils.data import TensorDataset, DataLoader

# Covert continuous features to a tensor
conts = np.stack([df[f].values for f in cont_features], 1)
conts = torch.tensor(conts, dtype=torch.float)

# Convert categorical variables to a tensor
cats = torch.tensor(cat_features, dtype=torch.int64) 

# Convert target features to a tensor
y = torch.tensor(df[out_features].values, dtype=torch.float).reshape(-1,1)


# # Get input data and target data
# data    = df[in_features].values
# target  = df[out_features].values

# # Create tensor using TensorDataset from the input and target features.
# cdm_td = TensorDataset(torch.FloatTensor(data), torch.FloatTensor(target))
# print(cdm_td[0])

In [ ]:
class LSTM(nn.Module):
    def __init__(self, input_size=1, hidden_size=50, out_size=1):
        super().__init__()
        self.hidden_size = hidden_size
        
        # Add an LSTM layer:
        self.lstm = nn.LSTM(input_size,hidden_size)
        
        # Add a fully-connected layer:
        self.linear = nn.Linear(hidden_size,out_size)
        
        # Initialize h0 and c0:
        self.hidden = (torch.zeros(1,1,hidden_size),
                       torch.zeros(1,1,hidden_size))
    
    def forward(self,seq):
        lstm_out, self.hidden = self.lstm(
            seq.view(len(seq), 1, -1), self.hidden)
        pred = self.linear(lstm_out.view(len(seq),-1))
        return pred[-1]   # we only care about the last prediction

In [ ]:
class TabularModel(nn.Module):

    def __init__(self, emb_szs, n_cont, out_sz, layers, p=0.5):
        
        # Inherit attributes from nn.Module class
        super().__init__()
        
        #############################################################################
        # Instanciate functions to use on the forward operation:
        
        # self.embeds: Creates a list of pre-configured Embedding operations (it is 
        # configured by passing the number of categories ni and the length of the 
        # embedding nf)
        self.embeds = nn.ModuleList([nn.Embedding(ni, nf) for ni,nf in emb_szs])
        
        # self.emb_drop: Cancels a proportion p of the embeddings.
        self.emb_drop = nn.Dropout(p)
        
        # self.bn_cont = Normalizes continuous features. This function is configured
        # by passing the number of continuous features to normalize.
        self.bn_cont = nn.BatchNorm1d(n_cont)
        
        #############################################################################
        # Count total number of embeddings (Total number of vector components for
        # every feature)
        n_emb = sum((nf for ni,nf in emb_szs))
        
        # Compute total number of inputs to pass to the initial layer (data point = 
        # Nb. of embeddings + Nb. of continuous variables)
        n_in = n_emb + n_cont
        
        # Run through every layer to set up the operations to perform per layer.
        # (i.e. layers=[100, 50, 200])
        layerlist = []
        for l, n_neurons in enumerate(layers):
            # On layer l, which contains n_neurons, perform the following operations:
            # 1. Apply linear neural network regression (z = Sum(wi*xi+bi))
            layerlist.append(nn.Linear(n_in,n_neurons))
            
            # 2. Apply ReLU activation function (al(z))
            layerlist.append(nn.ReLU(inplace=True))
            
            # 3. Normalize data using the n_neurons
            layerlist.append(nn.BatchNorm1d(n_neurons))
            
            # 4. Cancel out a random proportion p of the neurons to avoid overfitting
            layerlist.append(nn.Dropout(p))
            
            # 5. Set new number of input features n_in for the next layer l+1.
            n_in = n_neurons
        
        # Set the last layer of the list which corresponds to the final output
        layerlist.append(nn.Linear(layers[-1],out_sz))
        
        # Instantiate layers as a Neural Network sequential task
        self.layers = nn.Sequential(*layerlist)
    
    def forward(self, x_cat, x_cont):
        # Initialize embeddings list
        embeddings = []
        
        # Apply embedding function e from self.embeds to the category i
        # in x_cat array
        for i,e in enumerate(self.embeds):
            embeddings.append(e(x_cat[:,i]))
        
        # Concatenate embedding sections into 1
        x = torch.cat(embeddings, 1)
        
        # Apply dropout function to to the embeddings torch
        x = self.emb_drop(x)
        
        # Normalize continuous variables
        x_cont = self.bn_cont(x_cont)
        
        # Concatenate embeddings with continuous variables into one torch
        x = torch.cat([x, x_cont], 1)
        
        # Process all data points with the layers functions (sequential of operations)
        x = self.layers(x)
        
        return x


## Save the trained model to a file
Right now <strong><tt>model</tt></strong> has been trained and validated, and seems to correctly classify an iris 97% of the time. Let's save this to disk.<br>
The tools we'll use are <a href='https://pytorch.org/docs/stable/torch.html#torch.save'><strong><tt>torch.save()</tt></strong></a> and <a href='https://pytorch.org/docs/stable/torch.html#torch.load'><strong><tt>torch.load()</tt></strong></a><br>

There are two basic ways to save a model.<br>

The first saves/loads the `state_dict` (learned parameters) of the model, but not the model class. The syntax follows:<br>
<tt><strong>Save:</strong>&nbsp;torch.save(model.state_dict(), PATH)<br><br>
<strong>Load:</strong>&nbsp;model = TheModelClass(\*args, \*\*kwargs)<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;model.load_state_dict(torch.load(PATH))<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;model.eval()</tt>

The second saves the entire model including its class and parameters as a pickle file. Care must be taken if you want to load this into another notebook to make sure all the target data is brought in properly.<br>
<tt><strong>Save:</strong>&nbsp;torch.save(model, PATH)<br><br>
<strong>Load:</strong>&nbsp;model = torch.load(PATH))<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;model.eval()</tt>

In either method, you must call <tt>model.eval()</tt> to set dropout and batch normalization layers to evaluation mode before running inference. Failing to do this will yield inconsistent inference results.

For more information visit https://pytorch.org/tutorials/beginner/saving_loading_models.html